Metadata API: Rewrite comments

jku opened the pull request:
Try to keep dostrings and comments to the point, avoid mentioning
details if they are not necessary or are likely to become outdated
and try to minimize number of comment lines.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: Rewrite comments

Try to keep dostrings and comments to the point, avoid mentioning
details if they are not necessary or are likely to become outdated
and try to minimize number of comment lines.

Co-authored-by: Joshua Lock <jlock@vmware.com>
Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented on tuf/api/metadata.py:
```diff
@@ -891,18 +852,14 @@ class DelegatedRole(Role):
 
     Attributes:
         name: A string giving the name of the delegated role.
-        keyids: A set of strings each of which represents a given key.
-        threshold: An integer representing the required number of keys for that
-            particular role.
```
these were removed because they are Role attributes, not DelegatedRole attributes (and we don't document parent attributes in other classes)

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -46,10 +46,7 @@
-# with more 1000 lines, because all of the code here is logically connected
-# and currently, we are above 1000 lines by a small margin.
-# pylint: disable=C0302
+# pylint: disable=too-many-lines
```
Great, didn't know we can reference warning explicitly!

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -407,8 +399,9 @@ def bump_version(self) -> None:
+
     Please note that "Key" instances are not semanticly validated during
-    initialization. We consider this as responsibility of securesystemslib.
+    initialization: this only happens at signature verification time.
```
Do we plan to use the `Key` class besides signature verification?
If no other cases appear in mind then okay, let's be more precise, but if there are other cases we envision it could be better to be a little more abstract.

jku replied:
I can't think of anything else we would want to expose... we probably want to add a `Key.from_securesystemslib_key()` still but even there we likely will just assign the attributes and that's it

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -512,15 +504,15 @@ def verify_signature(
 class Role:
-    """A container class containing the set of keyids and threshold associated
-    with a particular role.
+    """Container for data that defines which keys are required to sign metadata
```
```suggestion
    """Container for data that defines which keys are required to sign metadata.
```

MVrachev reviewed (approved):
LGTM with the exception of one comment in the `Key` class.

jku commented:
Last push: Modified Role docstring based on review

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -512,15 +504,15 @@ def verify_signature(
-        threshold: An integer representing the required number of keys for that
-            particular role.
+        keyids: A set of strings representing signing keys for this role.
+        threshold: Number of keys required to sign this roles metadata.
```
```suggestion
        threshold: Number of keys required to sign this role's metadata.
```

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -786,13 +754,11 @@ def verify_length_and_hashes(self, data: Union[bytes, BinaryIO]):
+    Timestamp contains information about the snapshot Metadata file.
 
+    Attributes:
+        meta: A dictionary of filenames to MetaFiles. Only valid key value is
+            the snapshot filename, as defined by specification.
```
```suggestion
        meta: A dictionary of filenames to MetaFiles. The only valid key value
            is the snapshot filename, as defined by the specification.
```
IMNSHO these additional words increase ease of reading without increasing line count, but feel free to ignore.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -834,15 +800,10 @@ def update(self, snapshot_meta: MetaFile) -> None:
+    Snapshot contains information about all target Metadata files.
 
+    Attributes:
+        meta: A dictionary of target metadata filenames to MetaFiles.
```
```suggestion
        meta: A dictionary of target metadata filenames to MetaFile objects.
```
Wanted to preserve the keyword MetaFile

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -891,18 +852,14 @@ class DelegatedRole(Role):
         path_hash_prefixes: An optional list of HEX_DIGESTs used to succinctly
             describe a set of target paths. Only one of the attributes "paths"
-            and "path_hash_prefixes" is allowed to be set.
+            and "path_hash_prefixes" is allowed within a DelegatedRole.
```
I wonder if there's a clearer way to document that paths and paths_hash_prefixes are mutually exclusive?

jku replied:
I guess if we consider actual code changes we could just have a single list[str] attribute and then a way to say "does this DelegatedRole use hash prefixes Y/N?", but however you implement it (a flag attribute or even two separate classes) it it's going to look quite different compared to the file format ... so does not seem worth it

I think this might be worth pointing out in the body text of this class docstring... I'll try that

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -965,12 +922,11 @@ class Delegations:
+        keys: Dictionary of keyids to Keys. Defines the keys used in 'roles'.
+        roles: List of DelegatedRoles that define which keys are required to
+            sign the metadata for a specific role. The roles order also
+            defines the order the role delegations are considered.
```
```suggestion
            defines the order that role delegations are considered in.
```
?

I wanted to say "in which the role delegations are considered" but thought you might object to the character count increase.

joshuagl commented on tuf/api/metadata.py:
```diff
@@ -1082,25 +1030,18 @@ def verify_length_and_hashes(self, data: Union[bytes, BinaryIO]):
-        delegations: An optional object containing a list of delegated target
-            roles and public key store used to verify their metadata
-            signatures.
+    Targets contains verifying information about target files but also
+    delegates responsibility to other Targets files.
```
```suggestion
    Targets contains verifying information about target files and also
    delegates responsibility to other Targets roles.
```

joshuagl reviewed (commented):
Excellent work, thanks for doing this. ~60 fewer LOC and increases the clarity of the code comments.

I made a few minor suggestions, what do you think?

Commit:
Metadata: Improve DelegatedRole docstring

Explain the ways a delegation can happen: Do not try to cover the
complete process (specification should do that) but offer enough
details that the complexity is not completely hidden from the viewer.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
squashed joshuas suggestions (thanks!), then added another commit to improve the DelegatedRole docstring: It explains all of the options that exist in broad terms and expects reader to lookup the details (like what exactly is a "target path hash") elsewhere, and tries to make it clearer that the attributes are mutually exclusive.

joshuagl reviewed (approved):
LGTM, thanks!
